File: 0013. 罗马数字转整数.cpp
#include <unordered_map>

using namespace std;

class Solution {
 public:
  unordered_map<char, int> map = {
      // roman value
      {'I', 1},   {'V', 5},   {'X', 10},   {'L', 50},
      {'C', 100}, {'D', 500}, {'M', 1000},
  };
  int romanToInt(string s) {
    int result = 0;
    for (int i = 0; i < s.size(); i++) {
      int current_value = map[s[i]];
      if (map[s[i + 1]] > current_value) {
        current_value *= -1;
      }
      result += current_value;
    }
    return result;
  }
};

File: 0013. 罗马数字转整数.go
package leetcode

var romanCharValues = map[byte]int{'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}

func romanToInt(s string) int {
	result := 0
	for charIndex := range s {
		currentValue := romanCharValues[s[charIndex]]
		if charIndex < len(s)-1 && currentValue < romanCharValues[s[charIndex+1]] {
			result -= currentValue
		} else {
			result += currentValue
		}
	}
	return result
}
